src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/GMMVariantTrainDataManager.java

```diff
@@ -0,0 +1,443 @@
+        }
+        this.annotationKeys = new ArrayList<>( uniqueAnnotations );
+        this.GMMVTAC = GMMVTAC;
+        meanVector = new double[this.annotationKeys.size()];
+        varianceVector = new double[this.annotationKeys.size()];
+        trainingSets = new ArrayList<>();
+    }
+
+    public void setData( final List<VariantDatum> data ) {
+        this.data = data;
+    }
+
+    public void setNormalization(final Map<String, Double> anMeans, final Map<String, Double> anStdDevs) {
+        for (int i = 0; i < this.annotationKeys.size(); i++) {
+            meanVector[i] = anMeans.get(annotationKeys.get(i));
+            varianceVector[i] = anStdDevs.get(annotationKeys.get(i));
+        }
+    }
+
+    public List<VariantDatum> getData() {
+        return data;
+    }
+
+    /**
+     * Normalize annotations to mean 0 and standard deviation 1.
+     * Order the variant annotations by the provided list {@code theOrder} or standard deviation.
+     *
+     * @param calculateMeans Boolean indicating whether or not to calculate the means
+     * @param theOrder a list of integers specifying the desired annotation order. If this is null
+     *                 annotations will get sorted in decreasing size of their standard deviations.
+     */
+    public void normalizeData(final boolean calculateMeans, List<Integer> theOrder) {
+        boolean foundZeroVarianceAnnotation = false;
+        for( int iii = 0; iii < meanVector.length; iii++ ) {
+            final double theMean, theSTD;
+            if (calculateMeans) {
+                theMean = mean(iii, true);
+                theSTD = standardDeviation(theMean, iii, true);
+                if (Double.isNaN(theMean)) {
+                    throw new UserException.BadInput("Values for " + annotationKeys.get(iii) + " annotation not detected for ANY training variant in the input callset. VariantAnnotator may be used to add these annotations.");
+                }
+
+                foundZeroVarianceAnnotation = foundZeroVarianceAnnotation || (theSTD < 1E-5);
+                meanVector[iii] = theMean;
+                varianceVector[iii] = theSTD;
+            }
+            else {
+                theMean = meanVector[iii];
+                theSTD = varianceVector[iii];
+            }
+            logger.info(annotationKeys.get(iii) + String.format(": \t mean = %.2f\t standard deviation = %.2f", theMean, theSTD));
+            for( final VariantDatum datum : data ) {
+                // Transform each data point via: (x - mean) / standard deviation
+                datum.annotations[iii] = ( datum.isNull[iii] ? 0.1 * Utils.getRandomGenerator().nextGaussian() : ( datum.annotations[iii] - theMean ) / theSTD );
+            }
+        }
+        if( foundZeroVarianceAnnotation ) {
+            throw new UserException.BadInput( "Found annotations with zero variance. They must be excluded before proceeding." );
```

Comment:
I decided to WARN upon finding a zero-variance annotation and FAIL for any annotations that are completely missing across SNPs or INDELs. This check happens at the start of the training step and applies to any model backend, not just the GMM. I think this is better than failing earlier at the extract step, since extracted annotations are useful output regardless of whether they can be used to successfully train a model.

Note also that the new GMM and the IsolationForest should both be robust to zero-variance annotations---hence the WARN. Not sure if this will be true for all models, but it will be on them to check. I think it's reasonable to expect that models won't be able to handle completely missing annotations and fail in that case, though.